File: src/mapFaces.py
import csv
import os
from pathlib import Path
import shutil
import unicodedata
import logging

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(name)s - %(levelname)s - %(message)s",
    datefmt="%Y-%m-%d %H:%M:%S",
)
logger = logging.getLogger(__name__)

DELIMITER = ";"
ENCODING = "utf-8-sig"


class PlayerMapping:
    src_player_id: str
    dest_player_id: str

    def __init__(self, src_player_id: str, dest_player_id: str):
        self.src_player_id = src_player_id
        self.dest_player_id = dest_player_id


def read_csv(file_path: str):
    logger.info(f"Reading CSV file: {file_path}")
    data = {}
    try:
        with open(file_path, mode="r", encoding=ENCODING) as file:
            reader = csv.DictReader(file, delimiter=DELIMITER)
            for row in reader:
                player_id = row["Id"]
                player_name = row["Name"]
                data[player_name] = player_id
        logger.info(f"Successfully read {len(data)} players from {file_path}")
    except Exception as e:
        logger.error(f"Error reading CSV file {file_path}: {e}")
        raise
    return data


def get_player_mapping(source_csv: str, destination_csv: str) -> list[PlayerMapping]:
    logger.info("Starting player mapping process")
    source_data = read_csv(source_csv)
    destination_data = read_csv(destination_csv)
    player_mapping = []
    destination_names = list(destination_data.keys())

    for player_name in source_data.keys():
        candidate_name = get_best_match(player_name, destination_names)
        if candidate_name:
            player_mapping.append(
                PlayerMapping(
                    source_data[player_name], destination_data[candidate_name]
                )
            )
            logger.debug(f"Matched '{player_name}' -> '{candidate_name}'")
        else:
            logger.warning(f"No match found for player: {player_name}")

    logger.info(f"Successfully mapped {len(player_mapping)} players")
    return player_mapping


def update_faces_structure(
    src_folder_path: str, dest_folder_path: str, mapping: list[PlayerMapping]
):
    logger.info(
        f"Updating faces structure from {src_folder_path} to {dest_folder_path}"
    )
    facePath = "Asset/model/character/face/real"
    processed = 0

    for item in mapping:
        src_path = f"{src_folder_path}/{facePath}/{item.src_player_id}"
        dest_path = f"{dest_folder_path}/{facePath}/{item.dest_player_id}"

        if not os.path.exists(src_path):
            logger.warning(f"Source path does not exist: {src_path}")
            continue

        if not os.path.exists(dest_path):
            os.makedirs(dest_path)
            logger.debug(f"Created directory: {dest_path}")

        shutil.copytree(src_path, dest_path, dirs_exist_ok=True)
        hex_replace(
            f"{dest_path}/#Win/face.fpk", item.src_player_id, item.dest_player_id
        )
        processed += 1

    logger.info(f"Successfully processed {processed} player faces")


def hex_replace(file_path, old_id, new_id):
    if not Path(file_path).exists():
        logger.debug(f"File does not exist, skipping hex replace: {file_path}")
        return

    try:
        with open(file_path, "rb") as f:
            data = f.read()

        data = data.replace(old_id.encode(), new_id.encode())

        with open(file_path, "wb") as f:
            f.write(data)
        logger.debug(f"Hex replaced in {file_path}: {old_id} -> {new_id}")
    except Exception as e:
        logger.error(f"Error during hex replace in {file_path}: {e}")


def normalize(fullName: str):
    """Remove accents and non-alphanumeric characters (keep spaces)"""
    nfd = unicodedata.normalize("NFD", fullName)
    cleaned = "".join(
        character
        for character in nfd
        if unicodedata.category(character) != "Mn"
        and (character.isalnum() or character.isspace())
    )
    return cleaned.casefold()


def calculate_name_match_score(name1: str, name2: str):
    """Calculate match score. Higher is better. Returns None if no match."""
    parts1 = normalize(name1).split()
    parts2 = normalize(name2).split()

    if len(parts1) != len(parts2):
        return None

    if len(parts1) == 1:
        return 100 if parts1[0] == parts2[0] else None

    # Surname must match exactly
    if parts1[-1] != parts2[-1]:
        return None

    score = 0

    # Check all first names
    for p1, p2 in zip(parts1[:-1], parts2[:-1]):
        if len(p1) == 1 or len(p2) == 1:
            if p1[0] != p2[0]:
                return None
            score += 1  # Lower score for initial match
        else:
            if p1 != p2:
                return None
            score += 10  # Higher score for full name match

    score += 50  # Bonus for surname match
    return score


def get_best_match(target_name: str, candidate_names: list[str]):
    """Find best matching name from candidates. Returns closest match or None."""
    # Early return for exact match
    if target_name in candidate_names:
        return target_name

    best_match = None
    best_score = -1

    for candidate in candidate_names:
        score = calculate_name_match_score(target_name, candidate)
        if score is not None and score > best_score:
            best_score = score
            best_match = candidate

    return best_match


if __name__ == "__main__":
    logger.info("=== Starting Player Face Mapping Tool ===")

    source_csv = "samples/BPB-2023-players.csv"
    destination_csv = "samples/FL26_players.csv"
    src_folder_path = "samples"
    dest_folder_path = "livecpk/root"

    try:
        mapping = get_player_mapping(source_csv, destination_csv)
        update_faces_structure(src_folder_path, f"result/{dest_folder_path}", mapping)
        logger.info("=== Finished processing successfully ===")
    except Exception as e:
        logger.critical(f"Fatal error during processing: {e}", exc_info=True)
        raise

File: src/mapFaces.go
package main

import (
	"bytes"
	"encoding/csv"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"unicode"

	"golang.org/x/text/runes"
	"golang.org/x/text/transform"
	"golang.org/x/text/unicode/norm"
)

const (
	Delimiter = ';'
	Encoding  = "utf-8"
)

type PlayerMapping struct {
	SrcPlayerID  string
	DestPlayerID string
}

func readCSV(filePath string) (map[string]string, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	reader := csv.NewReader(file)
	reader.Comma = Delimiter
	reader.LazyQuotes = true

	records, err := reader.ReadAll()
	if err != nil {
		return nil, err
	}

	data := make(map[string]string)
	// Skip header row
	for i := 1; i < len(records); i++ {
		if len(records[i]) >= 2 {
			playerID := records[i][0]
			playerName := records[i][1]
			data[playerName] = playerID
		}
	}
	return data, nil
}

func normalize(fullName string) string {
	// NFD normalization
	t := transform.Chain(norm.NFD, runes.Remove(runes.In(unicode.Mn)), norm.NFC)
	result, _, _ := transform.String(t, fullName)

	// Keep only alphanumeric and spaces
	var builder strings.Builder
	for _, char := range result {
		if unicode.IsLetter(char) || unicode.IsNumber(char) || unicode.IsSpace(char) {
			builder.WriteRune(char)
		}
	}

	return strings.ToLower(builder.String())
}

func calculateNameMatchScore(name1, name2 string) *int {
	parts1 := strings.Fields(normalize(name1))
	parts2 := strings.Fields(normalize(name2))

	if len(parts1) != len(parts2) {
		return nil
	}

	if len(parts1) == 1 {
		if parts1[0] == parts2[0] {
			score := 100
			return &score
		}
		return nil
	}

	// Surname must match exactly
	if parts1[len(parts1)-1] != parts2[len(parts2)-1] {
		return nil
	}

	score := 0

	// Check all first names
	for i := 0; i < len(parts1)-1; i++ {
		p1 := parts1[i]
		p2 := parts2[i]

		if len(p1) == 1 || len(p2) == 1 {
			// Initial match
			if p1[0] != p2[0] {
				return nil
			}
			score += 1
		} else {
			// Full name match
			if p1 != p2 {
				return nil
			}
			score += 10
		}
	}

	score += 50 // Bonus for surname match
	return &score
}

func getBestMatch(targetName string, candidateNames []string) string {
	var bestMatch string
	bestScore := -1

	for _, candidate := range candidateNames {
		score := calculateNameMatchScore(targetName, candidate)
		if score != nil && *score > bestScore {
			bestScore = *score
			bestMatch = candidate
		}
	}

	return bestMatch
}

func getPlayerMapping(sourceCSV, destinationCSV string) ([]PlayerMapping, error) {
	sourceData, err := readCSV(sourceCSV)
	if err != nil {
		return nil, err
	}

	destinationData, err := readCSV(destinationCSV)
	if err != nil {
		return nil, err
	}

	var playerMapping []PlayerMapping
	destinationNames := make([]string, 0, len(destinationData))
	for name := range destinationData {
		destinationNames = append(destinationNames, name)
	}

	for playerName, srcID := range sourceData {
		candidateName := getBestMatch(playerName, destinationNames)
		if candidateName != "" {
			playerMapping = append(playerMapping, PlayerMapping{
				SrcPlayerID:  srcID,
				DestPlayerID: destinationData[candidateName],
			})
		}
	}

	return playerMapping, nil
}

func copyDir(src, dst string) error {
	return filepath.Walk(src, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}

		relPath, err := filepath.Rel(src, path)
		if err != nil {
			return err
		}
		dstPath := filepath.Join(dst, relPath)

		if info.IsDir() {
			return os.MkdirAll(dstPath, info.Mode())
		}

		data, err := os.ReadFile(path)
		if err != nil {
			return err
		}
		return os.WriteFile(dstPath, data, info.Mode())
	})
}

func hexReplace(filePath, oldID, newID string) error {
	if _, err := os.Stat(filePath); os.IsNotExist(err) {
		return nil
	}

	data, err := os.ReadFile(filePath)
	if err != nil {
		return err
	}

	data = bytes.ReplaceAll(data, []byte(oldID), []byte(newID))

	return os.WriteFile(filePath, data, 0644)
}

func updateFacesStructure(srcFolderPath, destFolderPath string, mapping []PlayerMapping) error {
	facePath := "Asset/model/character/face/real"

	for _, item := range mapping {
		srcPath := filepath.Join(srcFolderPath, facePath, item.SrcPlayerID)
		destPath := filepath.Join(destFolderPath, facePath, item.DestPlayerID)

		if _, err := os.Stat(srcPath); os.IsNotExist(err) {
			continue
		}

		if err := os.MkdirAll(destPath, 0755); err != nil {
			return err
		}

		if err := copyDir(srcPath, destPath); err != nil {
			return err
		}

		fpkPath := filepath.Join(destPath, "#Win", "face.fpk")
		if err := hexReplace(fpkPath, item.SrcPlayerID, item.DestPlayerID); err != nil {
			fmt.Printf("Warning: Could not replace hex in %s: %v\n", fpkPath, err)
		}
	}

	return nil
}

func main() {
	sourceCSV := "samples/BPB-2023-players.csv"
	destinationCSV := "samples/FL26_players.csv"
	srcFolderPath := "samples"
	destFolderPath := "livecpk/root"

	mapping, err := getPlayerMapping(sourceCSV, destinationCSV)
	if err != nil {
		fmt.Printf("Error getting player mapping: %v\n", err)
		os.Exit(1)
	}

	resultPath := filepath.Join("result", destFolderPath)
	if err := updateFacesStructure(srcFolderPath, resultPath, mapping); err != nil {
		fmt.Printf("Error updating faces structure: %v\n", err)
		os.Exit(1)
	}

	fmt.Println("Finished processing")
}
